ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,2157 @@
+    self.ret
+  }
+}
+
+impl From<&NativeType> for fast_api::Type {
+  fn from(native_type: &NativeType) -> Self {
+    match native_type {
+      NativeType::Bool => fast_api::Type::Bool,
+      NativeType::U8 | NativeType::U16 | NativeType::U32 => {
+        fast_api::Type::Uint32
+      }
+      NativeType::I8 | NativeType::I16 | NativeType::I32 => {
+        fast_api::Type::Int32
+      }
+      NativeType::F32 => fast_api::Type::Float32,
+      NativeType::F64 => fast_api::Type::Float64,
+      NativeType::Void => fast_api::Type::Void,
+      NativeType::I64 => fast_api::Type::Int64,
+      NativeType::U64 => fast_api::Type::Uint64,
+      NativeType::ISize => fast_api::Type::Int64,
+      NativeType::USize | NativeType::Pointer | NativeType::Function => {
+        fast_api::Type::Uint64
+      }
+      NativeType::Buffer => fast_api::Type::TypedArray(fast_api::CType::Uint8),
+    }
+  }
+}
+
+macro_rules! x64 {
+  ($assembler:expr; $($tokens:tt)+) => {
+    dynasm!($assembler; .arch x64; $($tokens)+)
+  }
+}
+
+macro_rules! aarch64 {
+  ($assembler:expr; $($tokens:tt)+) => {
+    dynasm!($assembler; .arch aarch64; $($tokens)+)
+  }
+}
+
+struct SysVAmd64 {
+  // Reference: https://refspecs.linuxfoundation.org/elf/x86_64-abi-0.99.pdf
+  assmblr: dynasmrt::x64::Assembler,
+  // Parameter counters
+  integer_params: u32,
+  float_params: u32,
+  // Stack offset accumulators
+  offset_trampoline: u32,
+  offset_callee: u32,
+  allocated_stack: u32,
+  frame_pointer: u32,
+}
+
+#[cfg_attr(
+  not(all(target_aarch = "x86_64", target_family = "unix")),
+  allow(dead_code)
+)]
+impl SysVAmd64 {
+  // Integer arguments go to the following GPR, in order: rdi, rsi, rdx, rcx, r8, r9
+  const INTEGER_REGISTERS: u32 = 6;
+  // SSE arguments go to the first 8 SSE registers: xmm0-xmm7
+  const FLOAT_REGISTERS: u32 = 8;
+
+  fn new() -> Self {
+    Self {
+      assmblr: dynasmrt::x64::Assembler::new().unwrap(),
+      integer_params: 0,
+      float_params: 0,
+      // Start at 8 to account for trampoline caller's return address
+      offset_trampoline: 8,
+      // default to tail-call mode. If a new stack frame is allocated this becomes 0
+      offset_callee: 8,
+      allocated_stack: 0,
+      frame_pointer: 0,
+    }
+  }
+
+  fn compile(sym: &Symbol) -> Trampoline {
+    let mut compiler = Self::new();
+
+    let must_cast_return_value =
+      compiler.must_cast_return_value(sym.result_type);
+    let must_wrap_return_value =
+      compiler.must_wrap_return_value_in_typed_array(sym.result_type);
+    let must_save_preserved_register = must_wrap_return_value;
+    let cannot_tailcall = must_cast_return_value || must_wrap_return_value;
+
+    if cannot_tailcall {
+      if must_save_preserved_register {
+        compiler.save_preserved_register_to_stack();
+      }
+      compiler.allocate_stack(&sym.parameter_types);
+    }
+
+    for param in &sym.parameter_types {
+      compiler.move_left(param)
+    }
```

Comment:
You don't have to shift if you leave in the receiver in the function prototype (zeroed out or not). In the simple case that means you can elide the trampoline completely.

I'm thinking of `fn square(_: *const (), x: u32) -> u32 { x * x }` - seems like a reasonable trade-off to me.

Reply:
Since this is FFI work, calling native library C APIs we can't just decide to leave the receiver in the function prototype. Sqlite's C API won't change for us even if we decide that it will.

Reply:
Okay, fair enough. I'm thinking of fast API calls into deno's runtime, where we control the function prototype. On the other hand, we probably wouldn't use FFI for that (or maybe...)

Reply:
Yeah, there's no reason to use FFI for those calls or to manually write assembly, as then we can just have Rust functions as the fast functions which will be even better optimized than our hand-rolled trampoline ASM.

Reply:
^ thats what `#[op]` is for...AOT fast api calls codegen for deno's runtime.